staging/src/k8s.io/apimachinery/pkg/runtime/converter.go

```diff
@@ -109,21 +110,131 @@ type unstructuredConverter struct {
 // to Go types via reflection. It performs mismatch detection automatically and is intended for use by external
 // test tools. Use DefaultUnstructuredConverter if you do not explicitly need mismatch detection.
 func NewTestUnstructuredConverter(comparison conversion.Equalities) UnstructuredConverter {
+	return NewTestUnstructuredConverterWithValidation(comparison)
+}
+
+// NewTestUnstrucutredConverterWithValidation allows for access to
+// FromUnstructuredWithValidation from within tests.
+func NewTestUnstructuredConverterWithValidation(comparison conversion.Equalities) *unstructuredConverter {
 	return &unstructuredConverter{
 		mismatchDetection: true,
 		comparison:        comparison,
 	}
 }
 
-// FromUnstructured converts an object from map[string]interface{} representation into a concrete type.
+const (
+	// FieldValidationIgnore ignores unknown/duplicate fields
+	FieldValidationIgnore = "Ignore"
+	// FieldValidationWarn responds with a warning, but successfully serve the request
+	FieldValidationWarn = "Warn"
+	// FieldValidationStrict fails the request on unknown/duplicate fields
+	FieldValidationStrict = "Strict"
+)
+
+var allowedValidationDirectives = sets.NewString(FieldValidationIgnore, FieldValidationWarn, FieldValidationStrict)
+
+// fromUnstructuredContext provides options for informing the converter
+// the state of its recursive walk through the conversion process.
+type fromUnstructuredContext struct {
+	// isInlined indicates whether the converter is currently in
+	// an inlined field or not to dtermine whether it should
+	// validate the matchedKeys yet or only collect them.
+	isInlined bool
+	// matchedKeys is the set of all fields that exist in the
+	// concrete go type of the object being converted into.
+	matchedKeys map[string]struct{}
+	// parentPath collects the path that the conversion
+	// takes as it travers the unstructured json map.
+	// It is used to report the full path to any unknown
+	// fields that the converter encounters.
+	parentPath *[]string
+	// previousParentLen indicates what the length of the
+	// parentPath previously was in the stack prior to the
+	// current recursive call and lets us reset back to it.
+	previousParentLen int
+	// validationDirective indicates what to do with unknown fields.
+	// It defaults to ignoring unknown fields. Options are:
+	// - FieldValidationIgnore
+	// - FieldValidationWarn
+	// - FieldValidationStrict
+	validationDirective string
+}
+
+func (c *fromUnstructuredContext) pushIndex(index int) {
+	if c.validationDirective == FieldValidationIgnore {
+		return
+	}
+	if c.parentPath == nil {
+		c.parentPath = &[]string{}
+	}
+	c.previousParentLen = len(*c.parentPath)
```

Comment:
Okay I think I've done all of this. The matched key tracking went pretty smoothly. My only question is if I am doing the path/isInlined tracking how you expected (ie. I reverted to what I had before and got rid of what wasn't needed because of the new matched key tracking)